backends/bmv2/synthesizeValidField.h

```diff
@@ -0,0 +1,92 @@
+/*
+Copyright 2013-present Barefoot Networks, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#ifndef _BACKENDS_BMV2_SYNTHESIZEVALIDFIELD_H_
+#define _BACKENDS_BMV2_SYNTHESIZEVALIDFIELD_H_
+
+#include "ir/pass_manager.h"
+
+namespace P4 {
+class ReferenceMap;
+class TypeMap;
+}  // namespace P4
+
+namespace BMV2 {
+
+/**
+ * Adds an explicit valid bit to each header type and replaces calls to
+ * isValid() with references to the new field. This better reflects how header
+ * validity is actually implemented on BMV2, which simplifies later passes.
+ *
+ * In most situations, `foo.isValid()` is rewritten to `foo.$valid$ == 1`.
+ * However, when `foo.isValid()` is a table key element (and isn't part of a
+ * larger expression), it's replaced with a simple match against `foo.$valid$`
+ * itself. This changes the type of the key element from boolean to bit<1>;
+ * to accomodate that change, the corresponding constant table entry keys are
+ * also rewritten to use bit<1> values instead of booleans.
+ *
+ * Calls to isValid() on header unions aren't rewritten; they can't be desugared
+ * to a simple read of a hidden field.
+ *
+ * XXX(seth): It would be cleaner to rewrite setValid() and setInvalid() too,
```

Comment:
I actually do think we're eventually going to be able to do this. It requires some BMV2 changes and they aren't in place yet, but the documentation (search for `$valid$` [here](https://github.com/p4lang/behavioral-model/blob/master/docs/JSON_format.md)) suggests they will probably arrive eventually.